forage-android/src/main/java/com/joinforage/forage/android/core/ui/element/ForagePinElement.kt

```diff
@@ -0,0 +1,186 @@
+                    val elementHeight: Int = getDimensionPixelSize(R.styleable.ForagePINEditText_elementHeight, ViewGroup.LayoutParams.WRAP_CONTENT)
+
+                    _linearLayout = LinearLayout(context)
+                    _linearLayout.layoutParams = ViewGroup.LayoutParams(elementWidth, elementHeight)
+                    _linearLayout.orientation = VERTICAL
+                    _linearLayout.gravity = Gravity.CENTER
+                } finally {
+                    recycle()
+                }
+            }
+    }
+
+    internal abstract fun determineBackingVault(forageConfig: ForageConfig, logger: Log): VaultWrapper
+
+    override fun initWithForageConfig(forageConfig: ForageConfig) {
+        // Must initialize DD at the beginning of each render function. DD requires the context,
+        // so we need to wait until a context is present to run initialization code. However,
+        // we have logging all over the SDK that relies on the render happening first.
+        val logger = Log.getInstance()
+        logger.initializeDD(context, forageConfig)
+
+        // defer to the concrete subclass for the details of obtaining
+        // and instantiating the backing vault instance. We just need
+        // to guarantee that the vault is instantiated prior to adding
+        // it to the parent view
+        _SET_ONLY_vault = determineBackingVault(forageConfig, logger)
+
+        _linearLayout.addView(vault.getTextElement())
+        _linearLayout.addView(getLogoImageViewLayout(context))
+        addView(_linearLayout)
+
+        logger.i("[View] ForagePINEditText successfully rendered")
+    }
+
+    override fun clearText() {
+        vault.clearText()
+    }
+
+    internal fun getVaultSubmitter(logger: Log): AbstractVaultSubmitter =
+        vault.getVaultSubmitter(this, logger)
+
+    // While the events that ForageElements expose mirrors the
+    // blur, focus, change etc events of an Android view,
+    // they represent different abstractions. Our users need to
+    // interact with the ForageElement abstraction and not the
+    // implementation details of which Android view we use.
+    // Therefore we expose novel set listener methods instead of
+    // overriding the convention setOn*Listener
+    override fun setOnFocusEventListener(l: SimpleElementListener) {
+        vault.setOnFocusEventListener(l)
+    }
+    override fun setOnBlurEventListener(l: SimpleElementListener) {
+        vault.setOnBlurEventListener(l)
+    }
+    override fun setOnChangeEventListener(l: StatefulElementListener<PinElementState>) {
+        vault.setOnChangeEventListener(l)
+    }
+
+    override fun getElementState(): PinElementState {
+        return vault.manager.getState()
+    }
+
+    internal fun getVaultType(): VaultType {
+        return vault.vaultType
+    }
+    internal fun getTextElement(): View {
+        return vault.getTextElement()
+    }
```

Comment:
We want `getTextElement()` to be capable of returning the input element for BT, VGS, and Rosetta inside of a VaultSubmitter. The only way for this the compiler to be OK with `getTextElement()` returning any of these inputs, is if we say that this method will return a common parent class of each of them. Because the implementations differ between BT, VGS, and Rosetta, the nearest common parent is the `View`.